logs_sdk/lib/opentelemetry/sdk/logs/logger.rb

```diff
@@ -25,6 +27,56 @@ def initialize(name, version, logger_provider)
           @instrumentation_scope = InstrumentationScope.new(name, version)
           @logger_provider = logger_provider
         end
+
+        def resource
+          logger_provider.resource
+        end
+
+        # Emit a {LogRecord} to the processing pipeline.
+        #
+        # @param timestamp [optional Float, Time] Time in nanoseconds since Unix
+        #   epoch when the event occurred measured by the origin clock, i.e. the
+        #   time at the source.
+        # @param observed_timestamp [optional Float, Time] Time in nanoseconds
+        #   since Unix epoch when the event was observed by the collection system.
+        #   Intended default: Process.clock_gettime(Process::CLOCK_REALTIME, :nanosecond)
+        # @param [optional OpenTelemetry::Trace::SpanContext] span_context The
+        #   OpenTelemetry::Trace::SpanContext to associate with the
+        #   {LogRecord}.
+        # @param severity_number [optional Integer] Numerical value of the
+        #   severity. Smaller numerical values correspond to less severe events
+        #   (such as debug events), larger numerical values correspond to more
+        #   severe events (such as errors and critical events).
+        # @param severity_text [optional String] Original string representation of
+        #   the severity as it is known at the source. Also known as log level.
+        # @param body [optional String, Numeric, Boolean, Array<String, Numeric,
+        #   Boolean>, Hash{String => String, Numeric, Boolean, Array<String,
+        #   Numeric, Boolean>}] A value containing the body of the log record.
+        # @param attributes [optional Hash{String => String, Numeric, Boolean,
+        #   Array<String, Numeric, Boolean>}] Additional information about the
+        #   event.
+        #
+        # @api public
+        def on_emit(timestamp: nil,
+                 observed_timestamp: nil,
+                 span_context: nil, # or should this just be context? like in the API?
+                 severity_number: nil,
+                 severity_text: nil,
+                 body: nil,
+                 attributes: nil)
+          log_record = LogRecord.new(timestamp: timestamp,
+                                     observed_timestamp: observed_timestamp,
+                                     span_context: span_context ||= OpenTelemetry::Trace.current_span.context,
+                                     severity_text: severity_text,
+                                     severity_number: severity_number,
+                                     body: body,
+                                     attributes: attributes,
+                                     logger: self)
+
+          logger_provider.instance_variable_get(:@log_record_processors).each do |processor|
+            processor.on_emit(log_record, span_context)
+          end
```

Comment:
I think this might not be the best implementation approach. I'd like some feedback here.

This made more sense when we were exposing `log_record_processors` with an `attr_reader`. However, we decided not to do that because [it might not be safe for concurrent execution.](https://github.com/open-telemetry/opentelemetry-ruby/pull/1517/commits/de7f36ce8fa8eab285dea579afe69e2501bfb635#r1346607101)

The [Span gets its processors defined as an attribute during creation](https://github.com/open-telemetry/opentelemetry-ruby/blob/98c629edbc6445305bed65c4f4306431d683b474/sdk/lib/opentelemetry/sdk/trace/tracer_provider.rb#L145), but it's also created by the TracerProvider rather than the Tracer. The Tracer is equivalent to the Logger in this scenario.

Also unlike Span, there isn't a `finish` operation separate from `start`, so [`emit` both creates the log record and sends it to the processing pipeline.](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord)

Reply:
That's a great idea, @xuan-cao-swi. Incorporated in 1f9f9027219888422d638fc4c24b6be15f52dd78 (along with other context-related updates).